Update SKIP 4 (import hierarchy, backporting, testing) (#8097)
scikit-image/scikit-image: doc/source/skips/4-transition-to-v2.rst

Implementation
--------------

This section documents concrete implementation details.

> As the implementation is realised and the consensus evolves, this section should be kept up to date to work as a reference.

Reviewer: ```suggestion
This section documents concrete implementation details, and will be updated as we progress with the v2 implementation.
```

First phase: Building `skimage2`
................................

Two new namespaces and Python packages, ``src/_skimage2`` and ``src/skimage2``, will be created alongside ``src/skimage``.

``_skimage2``

>   will be the temporary namespace, where we build up the new API.

Reviewer: ```suggestion
  will be the temporary namespace, where we build up the new API.
  It exists to help avoid circular imports during the transition phase,
  when `skimage` needs many of the v2 functions internally.
```

Author: That's only half of the truth. We are using `_skimage2` to avoid triggering the import warning in `skimage2`.

I'll reword along the spirit of your suggestion.

``skimage2``
  will import and expose the API in ``_skimage2`` to facilitate early testing downstream – nothing else should be included here.

>   It will be marked as *experimental* – importing it will warn that content in ``skimage2`` is still unstable.

Reviewer: ```suggestion
  is a lightweight wrapper around ``_skimage2``.
  It exposes the API for early testing and warns the user on import that 
  ``skimage2`` is still unstable.
```

> With the new namespaces available, we will start building the new API while preserving the existing one.

Reviewer: ```suggestion
The three-namespace-approach lets us build and experiment with the new API while preserving the existing `skimage` namespace and functionality.
This is an explicit goal to facilitate porting.
```

This process orients itself around the following principles:

Only one implementation
  If possible, only one implementation should exist, and one API should be a *simple* wrapper around the other.
  Implementations should preferably live in ``_skimage2``.

Import hierarchy
  ``skimage`` and ``skimage2`` may only import from ``_skimage2``, and ``_skimage2`` should be self-contained.

>   This import hierarchy avoids circular imports and triggering the *experimental* warning.

Reviewer: ```suggestion
  This import hierarchy avoids circular imports as well as triggering the *experimental* warning.
```

>   Exceptions should be temporary and should be realized as inlined imports.

Reviewer: ```suggestion
  Exceptions (for example, importing from `skimage` in `_skimage`) are temporary and should be realized using inlined imports.
```

  This allows reusing code in ``skimage`` from ``_skimage2``, that has not (yet) been ported.

Test coverage of both APIs
  Each API, especially the differences, should be covered by independent tests.

>   Behavior that is identical between both APIs may be covered by only one test suite, ideally covering the wrapping API.

Reviewer: ```suggestion
  The `skimage` test suite should be duplicated for `_skimage2` functions.
  The tests should be adjusted to verify new behavior.
  To counteract the slowdown due to test duplication, we will soon select tests
  using dependency analysis (see `#7749 <https://github.com/scikit-image/scikit-image/pull/7749>`__).
```

Implementation

> --------------

Author: @stefanv, I'd like to clarify the purpose of this section. Our [SKIP template](https://scikit-image.org/docs/dev/skips/template.html#implementation) describes this section as

> This section lists the major steps required to implement the SKIP. Where possible, it should be noted where one step is dependent on another, and which steps may be optionally omitted.

To me this implies that the section is almost like a specification and reference on what to do.

I think that's a good thing. My suggestion would be to add the background and reasons in a different section. But I feel like such a section – something like _Suppl_ – is missing from our [SKIP template](https://scikit-image.org/docs/dev/skips/template.html#implementation). _Discussion_ doesn't really fit from its description, neither does _Detailed description_.

What do you think about adding a _FAQ_ section like PEPs do ([example](https://peps.python.org/pep-0810/#faq))?

Author: I'll hold off on addressing your remaining suggestions, until I'm clear on where we want to put the reasoning.

Reviewer: Yes, I like that idea. Also PEPs often discuss alternatives considered, which I also like reading.

With the new namespaces available, we will start building the new API while preserving the existing ``skimage`` namespace and functionality.
This process orients itself around the following principles:

One implementation
  If possible, only one implementation should exist, and one API should be a *simple* wrapper around the other.

Import hierarchy
  ``skimage`` and ``skimage2`` may only import from ``_skimage2``, and ``_skimage2`` should be self-contained.
  Exceptions (for example, importing from ``skimage`` in ``_skimage2``) are temporary and should be realized using inlined imports.
  By the end of this phase, implementations in ``_skimage2`` should no longer rely on code in ``skimage``.

Test coverage of both APIs
  The ``skimage`` test suite should be duplicated for ``_skimage2`` functions.
  The tests should be adjusted to verify new behavior.
  To counteract the slowdown due to test duplication, we will soon select tests
  using dependency analysis (see `#7749 <https://github.com/scikit-image/scikit-image/pull/7749>`__).

Minimize API difference
  Keep the differences between the old and new API small to make the eventual transition easier for users.

>   Prefer conventional deprecations in the ``skimage`` namespace if possible.

Reviewer 2: Do we still think this is a good idea?  I think we've concluded that we should make all new deprecations happen in `_skimage2`.

Author: I think we've been flip-flopping on this somewhat. I myself am uncertain about a general policy. The few cases where this came up we have decided on a case by case basis. 

But for example in case of `corner_peaks` (https://github.com/scikit-image/scikit-image/issues/8057) we are doing exactly what you propose.

I'll reword this so that the sentence reflects your sentiment. But I'll keep the keyword "prefer" to indicate some flexibility on a case by case basis. Let me know in case that doesn't work for you.

Reviewer 3: I'd say that the deprecations SHOULD be in skimage 1.0.  Then the guidance would be to enable deprecation warnings, and address the warnings while porting.  We did this for `pymongo` 3.x -> 4.0.

Author: @blink1073 can you clarify what you mean by that? A deprecation warning will be in the `skimage` namespace in all cases. The question is will the eventual change be made in `skimage` or in `skimage2`.

Do you agree with the most recent state (after https://github.com/scikit-image/scikit-image/pull/8097/commits/3743f6be728d29967483214c8264bac644b07301)?

Reviewer 3: Okay, I see the distinction.  The analogy to `pymongo` would be to specify running your program with the `PendingSkimage2Change` warning enabled.

This process orients itself around the following principles:

One implementation
  If possible, only one implementation should exist, and one API should be a *simple* wrapper around the other.

.. _sk2-import-hierarchy:

Import hierarchy
  ``skimage`` and ``skimage2`` may only import from ``_skimage2``, and ``_skimage2`` should be self-contained.
  Exceptions (for example, importing from ``skimage`` in ``_skimage2``) are temporary and should be realized using inlined imports.
  By the end of this phase, implementations in ``_skimage2`` should no longer rely on code in ``skimage``.

Test coverage of both APIs
  The ``skimage`` test suite should be duplicated for ``_skimage2`` functions.
  The tests should be adjusted to verify new behavior.
  To counteract the slowdown due to test duplication, we will soon select tests
  using dependency analysis (see `#7749 <https://github.com/scikit-image/scikit-image/pull/7749>`__).

Minimize API difference
  Minimize the differences between the old and new API to make the eventual transition easier for users.

>   Conventional deprecation can still be introduced in ``skimage``, but deferring them to the ``_skimage2`` namespace is the preferred option.

Reviewer 2: ```suggestion
  Generally, any changes in behavior and associated deprecations should be
  done by a) making that change in ``_skimage2`` and b) pointing the user to
  the upcoming ``skimage2`` behavior.    We may also do conventional
  deprecation within ``skimage``,  if a strong argument can be made that there
  is an advantage to doing this.
```

Author: I realized that this is somewhat at odds with this principle's name _Minimize API difference_. Deferring changes to ski2 actually increases the difference with ski1. However, I guess this provides nuance and exceptions to this principle. I'll simplify your suggestion a little bit but otherwise will take you up on it.

Second phase: Transitioning to `skimage2`
.........................................

Once we consider the API in ``skimage2`` complete and stable, ``src/_skimage2`` will be merged into ``src/skimage2`` and the *experimental* warning will be removed.

Instead, we will mark the API in ``skimage`` as deprecated with a single top-level warning that is raised on import.
This warning will encourage users to transition to ``skimage2``.
It should link to the migration guide and should explain how to enable :ref:`more specific warnings <sk2-local-warning>`.

.. note::
   At this stage ``skimage2`` should no longer need to import from ``skimage`` to avoid triggering this new warning.

>    If this isn't the case, ``skimage`` needs to be split into a public and a private Python package that ``skimage2`` can (lazily) import from without triggering the new deprecation warning.

Reviewer: I'm not sure what this means.

Author: I brought this up during the last meeting. This is only relevant if we don't port every implementation to `skimage2`. Previously we left that door open for us but I think in the meeting most people agreed, that we will try to have every implementation in `skimage2`.

> This state will be published in a full release versioned 2.0.0.

Reviewer: ```suggestion
This state will be published in a full release as ``scikit-image==2.0.0``.
```

> From now on importing ``skimage2`` is encouraged.

Reviewer: ```suggestion
From that point onward, importing ``skimage2`` is encouraged.
```

Development of new features should only happen in ``skimage2``.
Bugs in ``skimage`` may still be addressed and will be included in releases with version >= 2.0.0.

> Bug fixes may be backported to a maintenance branch and made available as releases with versions <2.0.0, but this is not required and eventually discouraged.

Reviewer: ```suggestion
If we see fit, bug fixes may be backported to a maintenance branch and made available as releases with versions <2.0.0.
```

I don't see why, but fine to leave the door open ;)

> Not earlier than 1 year after the release of 2.0.0, we may begin to successively remove parts of the deprecated API from ``skimage``.

Reviewer: ```suggestion
No earlier than two years after the release of 2.0.0, we may start removing ``skimage`` – gradually in parts, or as a whole.
```

Implementations and internal code that still live in ``skimage`` will be moved to ``skimage2``.

> This process can and should be spread over multiple releases.

Reviewer: ```suggestion
```

Reviewer: An implementation detail we can debate later.

Second phase: Transitioning to `skimage2`
.........................................

Once we consider the API in ``skimage2`` complete and stable, ``src/_skimage2`` will be merged into ``src/skimage2`` and the *experimental* warning will be removed.

Instead, we will mark the API in ``skimage`` as deprecated with a single top-level warning that is raised on import.
This warning will encourage users to transition to ``skimage2``.
It should link to the migration guide and should explain how to enable :ref:`more specific warnings <sk2-local-warning>`.

.. note::
   At this stage ``skimage2`` should no longer need to import from ``skimage`` to avoid triggering this new warning.

This state will be published in a full release as ``scikit-image==2.0.0``.

From that point onward, importing ``skimage2`` is encouraged.
Development of new features should only happen in ``skimage2``.
Bugs in ``skimage`` may still be addressed and will be included in releases with version >= 2.0.0.
If we see fit, bug fixes may be backported to a maintenance branch and made available as releases with versions <2.0.0.

> No earlier than two years after the release of 2.0.0, we may start removing ``skimage`` – gradually in parts, or as a whole.

Reviewer 2: My conclusion from the discussion we had - was it on Zulip? - that we should a) not suggest the release of a maintenance branch of 1.x, but only release fixes on 2.x and b) we should not suggest gradual removal of `skimage` but removal as a whole, in due course (where "due course" may be many years from now).

Reviewer 3: My 2 cents - have a period of overlap, maybe for one minor version, where `skimage` and `skimage2` coexist, and then cut a branch for `1.x` and remove `skimage` from `main`.  `skimage` 1.x would then be in security fix mode.

Second phase: Transitioning to `skimage2`
.........................................

Once we consider the API in ``skimage2`` complete and stable, ``src/_skimage2`` will be merged into ``src/skimage2`` and the *experimental* warning will be removed.

Instead, we will mark the API in ``skimage`` as deprecated with a single top-level warning that is raised on import.
This warning will encourage users to transition to ``skimage2``.
It should link to the migration guide and should explain how to enable :ref:`more specific warnings <sk2-local-warning>`.

.. note::
   At this stage ``skimage2`` should no longer need to import from ``skimage`` to avoid triggering this new warning.

This state will be published in a full release as ``scikit-image==2.0.0``.

From that point onward, importing ``skimage2`` is encouraged.
Development of new features should only happen in ``skimage2``.
Bugs in ``skimage`` may still be addressed and will be included in releases with version >= 2.0.0.

> If we see fit, bug fixes may be backported to a maintenance branch and made available as releases with versions <2.0.0.

Reviewer 2: ```suggestion
```

I think we agreed we would not do this?

No earlier than two years after the ``skimage2`` namespace becomes public, we will remove the ``skimage`` namespace.

> We will provide at least 2 releases or 2 years (whichever is longer) of visible warnings.

Reviewer 2: ```suggestion
We will provide at least 2 releases or 2 years (whichever is longer) of visible warnings for every change necessary for porting from ``skimage`` to ``skimage2``.  That is, users will have at least 2 years after the warnings become visible to change their code, before their code will break after the latest update.
```

Author: Sounds good. Will merge this with the previous paragraph.

the ``skimage2`` release. Users will be notified about these resources - among
other things - by a warning in scikit-image 1.1.

Rationale
---------

Rationale for design decisions described in the previous section.
See *Alternatives* for why certain solutions where rejected.

Why so many namespaces?
.......................

Having additional namespaces significantly simplifies the gradual porting process during the first implementation phase.
The new API v2 can be gradually fleshed out in its own independent namespace and made available for testing to the community.

Separating ``_skimage2`` and ``skimage2`` simplifies the requirement that users are warned about the experimental status of the API v2.
Ported implementations that live in ``_skimage2`` can be reused by wrappers in ``skimage`` without triggering the import warning in ``skimage2``.

> An alternative would have been to raise this warning not on import of ``skimage2`` but when any of its attributes is used with a custom ``__getattr__``.

Reviewer 2: ```suggestion
An alternative would have been to raise this warning not on import of ``skimage2`` but when any of its attributes are used, with a custom ``__getattr__``.
```

However, we preferred to trigger this warning as early as possible.

Why restrict imports between namespaces?
........................................

Ported implementations in ``_skimage2`` may rely on functionality in ``skimage2`` that hasn't been ported yet.

> At the same time, wrappers in ``skimage`` may import from ``_skimage2``.

Reviewer 2: This wasn't clear to me.   Can you expand?
